mri/reconstruct/linear.py

```diff
@@ -114,13 +115,162 @@ def l2norm(self, shape):
         fake_data[tuple(zip(shape // 2))] = 1
 
         # Call mr_transform
         data = self.op(fake_data)
 
         # Compute the L2 norm
-        return numpy.linalg.norm(data)
+        return np.linalg.norm(data)
+
+
+class WaveletUD2(object):
+    """The wavelet undecimated operator using pysap wrapper.
+    """
+    def __init__(self, wavelet_id=24, nb_scale=4,
+                 verbose=0, multichannel=False):
+        """Init function for Undecimated wavelet transform
+
+        Parameters
+        -----------
+        wavelet_id: int, default 24 = undecimated (bi-) orthogonal transform
+            ID of wavelet being used
+        nb_scale: int, default 4
+            the number of scales in the decomposition.
+
+        Private Variables:
+            _has_run: Checks if the get_mr_filters was called already
+        """
+        self.wavelet_id = wavelet_id
+        self.multichannel = multichannel
+        self.nb_scale = nb_scale
+        self._opt = [
+            '-t{}'.format(self.wavelet_id),
+            '-n{}'.format(self.nb_scale),
+        ]
+        self._has_run = False
+        self.coeffs_shape = None
+        self.flatten = flatten
+        self.unflatten = unflatten
+
+    def _get_filters(self, shape):
+        """Function to get the Wavelet coefficients of Delta[0][0].
+        This function is called only once and later the
+        wavelet coefficients are obtained by convolving these coefficients
+        with input Data
+        """
+        self.transform = get_mr_filters(
+            tuple(shape),
+            opt=self._opt,
+            coarse=True,
+        )
+        self._has_run = True
+
+    def op(self, data):
+        """ Define the wavelet operator.
+
+        This method returns the input data convolved with the wavelet filter.
+
+        Parameters
+        ----------
+        data: ndarray or Image
+            input 2D data array.
+
+        Returns
+        -------
+        coeffs: ndarray
+            the wavelet coefficients.
+        """
+        if not self._has_run:
+            if self.multichannel:
+                self._get_filters(list(data.shape)[1:])
+            else:
+                self._get_filters(data.shape)
+        if self.multichannel:
+            coeffs = []
+            self.coeffs_shape = []
+            for channel in range(data.shape[0]):
+                coefs_real, coeffs_shape = self.flatten(
+                    filter_convolve(data[channel].real, self.transform))
+                coefs_imag, coeffs_shape = self.flatten(
+                    filter_convolve(data[channel].imag, self.transform))
+                coeffs.append(coefs_real + 1j * coefs_imag)
+                self.coeffs_shape.append(coeffs_shape)
+            return np.asarray(coeffs)
+        else:
+            coefs_real = filter_convolve(data.real, self.transform)
+            coefs_imag = filter_convolve(data.imag, self.transform)
+            coeffs, self.coeffs_shape = self.flatten(
```

Comment:
Why would we need to flatten the coefficients?

Reply:
Well, going forward, most of the proximity operators (especially LASSO, OWL etc, rely on this type of formatting)  I am merely ensuring consistency in the way `WaveletN` and `WaveletUD2` classes work

Reply:
I guess now is not really the time to talk about it nor the place, but I think it would be nicer if it was the proximity operators job to perform the flattening.

Reply:
Yes, I completely agree with you. All I am saying is that we would have wider impact with it, and if we plan to do it, we need to do it everywhere (ie waveletN also). For now, do you agree that we need consistency here?

Reply:
Yes like I said it's not the time or place (that's why I marked the conversation as solved)